SPECIFICATION.md

```diff
@@ -1,198 +1,143 @@
+* For endpoints that return a `200 OK` but have an error in the body, we need to check for these on each response. This also includes partial OKs, like with the SMS API. These should be exposed as an error to the user.
+* Although most of Nexmo's APIs are mostly RESTful, this is an implementation detail, and not something we should necessarily promote to the user, who should not care about the semantics of different HTTP verbs such as GET, POST or PUT.
+* Errors should be reported using the language's error functionality. In languages like Java and Python, this would involve throwing an appropriate Exception object. In languages such as Go or Rust, this would involve returning an appropriate error from the function. This means any object returned from the API which represents an error should be used to populate an error type.
+* We should keep third-party dependencies lightweight, but we should not shy away from bringing in third-party dependencies where they make sense. Where depending on third party libraries we should use libraries that are 'de-facto standard' wherever possible. No proprietary dependencies on the SDKs to minimize dependencies disappearing due to commercial interests.
+  
+# Callbacks
+Support should be provided for parsing callback data and verifying the callback. Details to be hashed out.
+Callback support may be provided in a separate module with extra dependencies to the core nexmo library.
+
+# Behavior
+## HTTP Transports
+The libraries should all make use of each ecosystem's best-quality HTTP library, even if it means an added 3rd-party dependency. Each client object should maintain a pool of open HTTP connections, making subsequent calls less expensive.
+
+## Logging
+Each SDK should use the standardized logging system as determined by the language (a PSR-3 client like Monolog for PHP, Log4J for Java, etc). The SDKs should have DEBUG-level logging which can be turned on and off by the users. We should redact client secrets. 
+
+## Pagination
+By default, pagination should be hidden from the user and exposed as a cursor or iterable for high-level functions. Users may, if they elect, drop down to a closer API level client to do manual paging. For example, a library will expose a higher-level function to retrieve objects that will do proper hydration and dependency injection, while allowing the user to use a direct API object to get back raw values.
+
+The higher level iterable/cursors should be smart enough to reduce the number of HTTP requests that they make with sane defaults. For example, a paging size of 1 or 100 can be suboptimal, where a paging size of 5 or 10 may be more suitable. This value may change based on the product/endpoint.
+
+Examples:
+* `$client->conversations()->searchConversations($filter)` will return an iterable that returns Conversation objects
+* `$client->conversations()->getApiClient()->get($filter, 3)` will return page 3 of the given filter results
```

Comment:
This is only possible when using page based navigation (which `conversations` doesn't use). Should we update the example to use number search (or similar)?

Reply:
Updated the text to handle cursor searching, and adjusted the examples